support for safe timers, periodic scheduling, #16742

patriknw opened the pull request:
Plenty of FIXME but worth early review.

Refs #16742

patriknw committed:
support for safe timers, periodic scheduling, #16742

* implemented an Intercept behavior that can handle the
  scheduled TimerMsg, even though it's of a different type
* Intercept is similar to Tap but more powerful, implemented
  Tap with Intercept. Intercept is internal API.
* When wrapping a behavior, e.g. Tap, the outer behavior must also be
  Deferred if the wrapped behavior is Deferred
* PostStop not signaled when stopped voluntarily, intercept messages
  to cancel timers when stopped is returned

johanandren commented on akka-typed-tests/src/test/scala/akka/typed/TimerSpec.scala:
```diff
@@ -0,0 +1,148 @@
+
+    def `must schedule repeated ticks`(): Unit = {
+      val probe = TestProbe[Event]("evt")
+      val behv = timerScheduler[Command] { timer ⇒
```
The other behaviours is something the actor is `deferred`, `immutable`, `mutable` while this is something it gets injected. Should it perhaps be `timed` instead, or `withTimer`?

patriknw replied:
I like `withTimer`

patriknw commented:
Pushed new implementation that is not based on spawnAdapter.
* spawnAdapter function is invoked synchronously on caller's thread,
  so it can't be used for this
* implemented an Intercept behavior that can handle the
  scheduled TimerMsg, even though it's of a different type
* Intercept is similar to Tap but more powerful, implemented
  Tap with Intercept. Intercept is internal API.

patriknw commented:
I think this is complete now, ready for final review.

ktoso commented on akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala:
```diff
@@ -0,0 +1,139 @@
+  final case class Timer[T](key: Any, msg: T, repeat: Boolean, generation: Int, task: Cancellable)
+  final case class TimerMsg(key: Any, generation: Int, owner: AnyRef)
+
+  def withTimer[T](factory: TimerSchedulerImpl[T] ⇒ Behavior[T]): Behavior[T] = {
```
Ah, sneaky - very good type juggling so we just need to implement it once but is exposed using the right types to each of the DSLs. I also like the `withXxx` style for all kinds of additional features an actor may want 👍

ktoso commented on akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala:
```diff
@@ -0,0 +1,139 @@
+      beforeOnMessage = interceptTimerMsg,
+      beforeOnSignal = (ctx, sig) ⇒ {
+      sig match {
+        case PreRestart | PostStop ⇒ cancelAll()
```
Hah! Very nice, biggest improvements of the entire PR to timers in actors in general I think :)

ktoso commented on akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala:
```diff
@@ -0,0 +1,139 @@
+
+  override def cancel(key: Any): Unit = {
+    timers.get(key) match {
+      case None    ⇒
```
comment that `// do nothing` is intentional?

ktoso replied:
or perhaps it should log when trying to cancel already cancelled one actually?

ktoso commented on akka-typed-tests/src/test/java/akka/typed/javadsl/ActorCompile.java:
```diff
@@ -56,6 +59,13 @@ public MyMsgB(String greeting) {
   }
 
+  {
+    Behavior<MyMsg> b = Actor.withTimer(timers -> {
```
naming nitpick: perhaps this is `withTimers` as it does not start a timer, but give you the capability to do so?

patriknw replied:
good suggestion

ktoso commented on akka-typed-tests/src/test/scala/akka/typed/TimerSpec.scala:
```diff
@@ -0,0 +1,207 @@
+    def target(monitor: ActorRef[Event], timer: TimerScheduler[Command], bumpCount: Int): Behavior[Command] = {
+      def bump(): Behavior[Command] = {
+        val nextCount = bumpCount + 1
+        timer.startTimer("T", Tick(nextCount), interval, repeat = true)
```
the API with `repeat = true` looks good in scala however in java it's IMHO a bit confusing to have a boolean in there - always one more guessing step what the boolean means. Sure we don't want to have startPeriodicTimer or other ways to have a dedicated name for them?

Maybe `timer.scheduleOnce` / `timer.schedulePeriodic`? The issue with current scheduler names is that "`schedule`" is periodic, but does not immediately look like it is, so sometimes people do that and don't realise they did a periodic one I thought?

patriknw replied:
I like that, changed to `startPeriodicTimer` and `startSingleTimer`

ktoso commented on akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala:
```diff
@@ -0,0 +1,41 @@
+ * It is used with `Actor.withTimer`, which also takes care of the
+ * lifecycle of the timers such as cancelling them when the actor
+ * is restarted or stopped.
+ */
```
Rephrase to make it less hidden that on a restart a timer must be scheduled again?

```scala
/**
 * Support for scheduled `self` messages in an actor.
 * It is used with `Actor.withTimer`. 
 * Timers are bound to the lifecycle of the actor that owns it, 
 * and thus are cancelled automatically when it is restarted or stopped.
 */
```

?

ktoso commented on akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala:
```diff
@@ -0,0 +1,41 @@
+  def isTimerActive(key: Any): Boolean
+
+  /**
+   * Cancel a timer with a given `key`. It's guaranteed that a message from the
```
I'd rephrase the guarantee a bit, people may not be sure what we mean by previous here.

```
Cancel a timer with a given `key`. 
If canceling a timer that was already canceled, or key never was used to start a timer this operation is a no-op.

It is guaranteed that a message from a canceled timer, including its previous incarnation for the same key,
will not be received by the actor, even though the message might already be enqueued in the mailbox
when cancel is called.
```

?

patriknw replied:
good doc improvements

ktoso commented on akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala:
```diff
@@ -0,0 +1,41 @@
+  def startTimer(key: Any, msg: T, timeout: FiniteDuration, repeat: Boolean): Unit
+
+  /**
+   * Check if a timer with a given `key` is started.
```
is started => is active

ktoso commented on akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala:
```diff
@@ -0,0 +1,41 @@
+   * Each timer has a key and if a new timer with same key is started
+   * the previous is cancelled and it's guaranteed that a message from the
+   * previous timer is not received, even though it might already be enqueued
+   * in the mailbox when the new timer is started.
```
good semantics 👍

ktoso commented on akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala:
```diff
@@ -210,6 +211,15 @@ object Actor {
+  /**
+   * Support for scheduled `self` messages in an actor.
+   * It takes care of the lifecycle of the timers such as cancelling them when the actor
+   * is restarted or stopped.
```
👍

ktoso reviewed (approved):
Some small comments, looking great though 👍  
LGTM

patriknw commented:
Pushed the fixes for Deferred and PostStop, and incorporated feedback

johanandren reviewed (commented):
LGTM
